docs/dev/stream/checkpointing.md

```diff
@@ -124,8 +124,8 @@ env.getCheckpointConfig.setMaxConcurrentCheckpoints(1)
 
 ## Selecting a State Backend
 
-The checkpointing mechanism stores the progress in the data sources and data sinks, the state of windows, as well as the [user-defined state](state.html) consistently to
-provide *exactly once* processing semantics. Where the checkpoints are stored (e.g., JobManager memory, file system, database) depends on the configured
+The checkpointing mechanism stores consistent snapshots of the progress in the data sources and data sinks, the state of windows and timers, as well as any [user-defined state](state.html), thereby
```

Comment:
Perhaps this first sentence could use a little more work. Can we better explain "progress"? "as well as" -> "and"? Could we move "providing *exactly once* processing semantics" to the beginning: "The checkpointing mechanism provides ..."?